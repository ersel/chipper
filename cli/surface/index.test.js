const fs = require('fs');

const surfaceAction = require('./');
const scanner = require('../../lib/scanner/cache/scanCache');
const openFile = require('../utils/openFile/');

jest.mock('../utils/openFile/');
jest.mock('fs');
jest.mock('../../lib/scanner/cache/scanCache', () =>
	jest.fn(() => {
		const PROJECT_ROOT = process.cwd();
		const MOCK_SCAN_DATA = [
			{
				sourceFile: `${PROJECT_ROOT}/src/documentsService.js`,
				importedModules: [
					{
						source: 'lodash',
						absolute: `${PROJECT_ROOT}/node_modules/lodash/lodash.js`,
						imports: [
							{
								imported: 'uniq',
								local: 'uniq'
							}
						]
					}
				]
			},
			{
				sourceFile: `${PROJECT_ROOT}/src/loginService.js`,
				importedModules: [
					{
						source: './documentsService',
						absolute: `${PROJECT_ROOT}/src/documentsService.js`,
						imports: [
							{
								imported: 'getDocuments',
								local: 'getLoggedInUsersDocuments'
							}
						]
					},
					{
						source: './utils/cookiesService',
						absolute: `${PROJECT_ROOT}/src/utils/cookiesService.js`,
						imports: [
							{
								imported: 'setCookie',
								local: 'setCookie'
							}
						]
					}
				]
			}
		];
		return Promise.resolve(MOCK_SCAN_DATA);
	})
);

describe('surface command', () => {
	let testOptions = {};
	beforeEach(() => {
		testOptions = {
			targetDir: '.',
			projectRoot: process.cwd(),
			ext: ['js'],
			incl: ['**/*.js'],
			excl: ['node_modules/**'],
			alias: 'some-alias:/path/to/some-alias',
			fileScanParallelism: 50,
			rescan: true
		};

		openFile.mockReset();
		fs.writeFileSync.mockReset();
	});
	it('should call scanner with the right params', done => {
		surfaceAction({ target: 'lodash' }, testOptions).then(() => {
			expect(scanner).lastCalledWith(
				{
					aliases: { 'some-alias': '/path/to/some-alias' },
					excludedPatterns: ['node_modules/**'],
					extensions: ['js'],
					fileScanParallelism: 50,
					includedPatterns: ['**/*.js'],
					projectRootPath: process.cwd(),
					targetDirectory: '.'
				},
				testOptions.rescan
			);
			done();
		});
	});

	it('should call scanner with the right params when no aliases are defined', done => {
		delete testOptions.alias;
		surfaceAction({ target: 'lodash' }, testOptions).then(() => {
			expect(scanner).lastCalledWith(
				{
					aliases: {},
					excludedPatterns: ['node_modules/**'],
					extensions: ['js'],
					fileScanParallelism: 50,
					includedPatterns: ['**/*.js'],
					projectRootPath: process.cwd(),
					targetDirectory: '.'
				},
				testOptions.rescan
			);
			done();
		});
	});
	it('should find node_module imports', done => {
		fs.statSync = () => true; // when set to true gets treated as node_modules import
		surfaceAction({ target: 'lodash' }, testOptions).then(() => {
			expect(fs.writeFileSync.mock.calls).toHaveLength(1);
			const lastReportWriteCall = fs.writeFileSync.mock.calls[0];
			const [fileName, reportContents] = lastReportWriteCall;
			expect(reportContents).toMatchSnapshot();
			expect(openFile).lastCalledWith(fileName);
			expect(openFile).toHaveBeenCalledTimes(1);
			done();
		});
	});

	it('should find local imports (relative path)', done => {
		fs.statSync = () => false;
		surfaceAction({ target: 'src/documentsService.js' }, testOptions).then(
			() => {
				expect(fs.writeFileSync.mock.calls).toHaveLength(1);
				const lastReportWriteCall = fs.writeFileSync.mock.calls[0];
				const [fileName, reportContents] = lastReportWriteCall;
				expect(reportContents).toMatchSnapshot();
				expect(openFile).lastCalledWith(fileName);
				expect(openFile).toHaveBeenCalledTimes(1);
				done();
			}
		);
	});

	it('should find local imports (absolute path)', done => {
		fs.statSync = () => false;
		surfaceAction(
			{ target: `${process.cwd()}/src/documentsService.js` },
			testOptions
		).then(() => {
			expect(fs.writeFileSync.mock.calls).toHaveLength(1);
			const lastReportWriteCall = fs.writeFileSync.mock.calls[0];
			const [fileName, reportContents] = lastReportWriteCall;
			expect(reportContents).toMatchSnapshot();
			expect(openFile).lastCalledWith(fileName);
			expect(openFile).toHaveBeenCalledTimes(1);
			done();
		});
	});

	it('should find aliased imports', done => {
		fs.statSync = () => false;
		testOptions.alias = `utils:${process.cwd()}/src/utils`;
		surfaceAction({ target: 'utils' }, testOptions).then(() => {
			expect(fs.writeFileSync.mock.calls).toHaveLength(1);
			const lastReportWriteCall = fs.writeFileSync.mock.calls[0];
			const [fileName, reportContents] = lastReportWriteCall;
			expect(reportContents).toMatchSnapshot();
			expect(openFile).lastCalledWith(fileName);
			expect(openFile).toHaveBeenCalledTimes(1);
			done();
		});
	});
});
