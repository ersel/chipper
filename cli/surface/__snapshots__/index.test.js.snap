// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`surface command should find aliased imports 1`] = `
"
	<html>
	<head>
	<title>Chipper Scan Results</title>
	<style>
	
	table.blueTable {
		border: 1px solid #1C6EA4;
		background-color: #EEEEEE;
		width: 100%;
		text-align: left;
		border-collapse: collapse;
	  }
	  table.blueTable td, table.blueTable th {
		border: 1px solid #AAAAAA;
		padding: 3px 2px;
	  }
	  table.blueTable tbody td {
		font-size: 18px;
	  }
	  table.blueTable tr:nth-child(even) {
		background: #D0E4F5;
	  }
	  table.blueTable thead {
		background: #1C6EA4;
		background: -moz-linear-gradient(top, #5592bb 0%, #327cad 66%, #1C6EA4 100%);
		background: -webkit-linear-gradient(top, #5592bb 0%, #327cad 66%, #1C6EA4 100%);
		background: linear-gradient(to bottom, #5592bb 0%, #327cad 66%, #1C6EA4 100%);
		border-bottom: 2px solid #444444;
	  }
	  table.blueTable thead th {
		font-size: 15px;
		font-weight: bold;
		color: #FFFFFF;
		border-left: 2px solid #D0E4F5;
	  }
	  table.blueTable thead th:first-child {
		border-left: none;
	  }
	
	</style>
	</head>
	<table class=\\"blueTable\\">
	<thead>
	<tr>
	<th>Source Module</th><th>Imported Module</th><th>Imported Symbols</th></tr></thead><tr><td>/src/loginService.js</td><td>/src/utils/cookiesService.js</td><td>setCookie</td></tr></table></html>"
`;

exports[`surface command should find local imports (absolute path) 1`] = `
"
	<html>
	<head>
	<title>Chipper Scan Results</title>
	<style>
	
	table.blueTable {
		border: 1px solid #1C6EA4;
		background-color: #EEEEEE;
		width: 100%;
		text-align: left;
		border-collapse: collapse;
	  }
	  table.blueTable td, table.blueTable th {
		border: 1px solid #AAAAAA;
		padding: 3px 2px;
	  }
	  table.blueTable tbody td {
		font-size: 18px;
	  }
	  table.blueTable tr:nth-child(even) {
		background: #D0E4F5;
	  }
	  table.blueTable thead {
		background: #1C6EA4;
		background: -moz-linear-gradient(top, #5592bb 0%, #327cad 66%, #1C6EA4 100%);
		background: -webkit-linear-gradient(top, #5592bb 0%, #327cad 66%, #1C6EA4 100%);
		background: linear-gradient(to bottom, #5592bb 0%, #327cad 66%, #1C6EA4 100%);
		border-bottom: 2px solid #444444;
	  }
	  table.blueTable thead th {
		font-size: 15px;
		font-weight: bold;
		color: #FFFFFF;
		border-left: 2px solid #D0E4F5;
	  }
	  table.blueTable thead th:first-child {
		border-left: none;
	  }
	
	</style>
	</head>
	<table class=\\"blueTable\\">
	<thead>
	<tr>
	<th>Source Module</th><th>Imported Module</th><th>Imported Symbols</th></tr></thead><tr><td>/src/loginService.js</td><td>/src/documentsService.js</td><td>getDocuments as getLoggedInUsersDocuments</td></tr></table></html>"
`;

exports[`surface command should find local imports (relative path) 1`] = `
"
	<html>
	<head>
	<title>Chipper Scan Results</title>
	<style>
	
	table.blueTable {
		border: 1px solid #1C6EA4;
		background-color: #EEEEEE;
		width: 100%;
		text-align: left;
		border-collapse: collapse;
	  }
	  table.blueTable td, table.blueTable th {
		border: 1px solid #AAAAAA;
		padding: 3px 2px;
	  }
	  table.blueTable tbody td {
		font-size: 18px;
	  }
	  table.blueTable tr:nth-child(even) {
		background: #D0E4F5;
	  }
	  table.blueTable thead {
		background: #1C6EA4;
		background: -moz-linear-gradient(top, #5592bb 0%, #327cad 66%, #1C6EA4 100%);
		background: -webkit-linear-gradient(top, #5592bb 0%, #327cad 66%, #1C6EA4 100%);
		background: linear-gradient(to bottom, #5592bb 0%, #327cad 66%, #1C6EA4 100%);
		border-bottom: 2px solid #444444;
	  }
	  table.blueTable thead th {
		font-size: 15px;
		font-weight: bold;
		color: #FFFFFF;
		border-left: 2px solid #D0E4F5;
	  }
	  table.blueTable thead th:first-child {
		border-left: none;
	  }
	
	</style>
	</head>
	<table class=\\"blueTable\\">
	<thead>
	<tr>
	<th>Source Module</th><th>Imported Module</th><th>Imported Symbols</th></tr></thead><tr><td>/src/loginService.js</td><td>/src/documentsService.js</td><td>getDocuments as getLoggedInUsersDocuments</td></tr></table></html>"
`;

exports[`surface command should find node_module imports 1`] = `
"
	<html>
	<head>
	<title>Chipper Scan Results</title>
	<style>
	
	table.blueTable {
		border: 1px solid #1C6EA4;
		background-color: #EEEEEE;
		width: 100%;
		text-align: left;
		border-collapse: collapse;
	  }
	  table.blueTable td, table.blueTable th {
		border: 1px solid #AAAAAA;
		padding: 3px 2px;
	  }
	  table.blueTable tbody td {
		font-size: 18px;
	  }
	  table.blueTable tr:nth-child(even) {
		background: #D0E4F5;
	  }
	  table.blueTable thead {
		background: #1C6EA4;
		background: -moz-linear-gradient(top, #5592bb 0%, #327cad 66%, #1C6EA4 100%);
		background: -webkit-linear-gradient(top, #5592bb 0%, #327cad 66%, #1C6EA4 100%);
		background: linear-gradient(to bottom, #5592bb 0%, #327cad 66%, #1C6EA4 100%);
		border-bottom: 2px solid #444444;
	  }
	  table.blueTable thead th {
		font-size: 15px;
		font-weight: bold;
		color: #FFFFFF;
		border-left: 2px solid #D0E4F5;
	  }
	  table.blueTable thead th:first-child {
		border-left: none;
	  }
	
	</style>
	</head>
	<table class=\\"blueTable\\">
	<thead>
	<tr>
	<th>Source Module</th><th>Imported Module</th><th>Imported Symbols</th></tr></thead><tr><td>/src/documentsService.js</td><td>/node_modules/lodash/lodash.js</td><td>uniq</td></tr></table></html>"
`;
